### DISCUSSION TITLE
etcd authentication

### DISCUSSION THREAD

**Discussion by oison111** [etcd-io/etcd#14298] [Q&A]

**Comment by oison111** [comment]
![image](https://user-images.githubusercontent.com/102855127/182425437-4c342821-2ae9-4200-b2e7-55ebc2cd81de.png)

The certificate is configured.

HTTP or HTTPS here ????

**Comment by oison111** [comment]
i have done this 
![image](https://user-images.githubusercontent.com/102855127/182433605-2584af12-7b80-4763-ae8a-65aa7f81f30e.png)
![image](https://user-images.githubusercontent.com/102855127/182434097-b653466d-27c7-464a-950f-599f8b57e942.png)
![image](https://user-images.githubusercontent.com/102855127/182434177-60639d24-d340-4d03-88e0-6d2db106ab2f.png)


![image](https://user-images.githubusercontent.com/102855127/182433334-651422ae-90fc-4935-ad17-aa59f4eeb5a5.png)

why
![image](https://user-images.githubusercontent.com/102855127/182433512-f7098469-17fb-4b49-a241-5f5076aca7b8.png)

**Comment by SimFG** [comment]
There is a problem with the way of writing the yml file. you can refer to `etcd.conf.yml.sample`
https://github.com/etcd-io/etcd/blob/main/etcd.conf.yml.sample

**Reply by oison111** [reply]
> There is a problem with the way of writing the yml file. you can refer to `etcd.conf.yml.sample` https://github.com/etcd-io/etcd/blob/main/etcd.conf.yml.sample

![image](https://user-images.githubusercontent.com/102855127/182560492-c57a056f-4c83-483d-9e0d-e83a7d93c43d.png)


I've already revised it；Run again and report another error

![image](https://user-images.githubusercontent.com/102855127/182557811-a365e020-ad6e-42c7-89ad-be1bfde8c6e1.png)

**Reply by SimFG** [reply]
maybe you need to check some tls files. The tls files has been identified.

**Reply by SimFG** [reply]
refer to: https://etcd.io/docs/v3.5/op-guide/security/
example 3, maybe you can try it.

**Reply by oison111** [reply]
> refer to: https://etcd.io/docs/v3.5/op-guide/security/ example 3, maybe you can try it.

![image](https://user-images.githubusercontent.com/102855127/182587042-8143590d-f2b7-415c-8265-925c3853ccd5.png)

When I generated certificates，I only generated one set of certificates for both server and client. The hosts field was left blank because I didn't want to limit the number of hosts.

Can I do this ？？？

**Reply by oison111** [reply]
> refer to: https://etcd.io/docs/v3.5/op-guide/security/ example 3, maybe you can try it.

![image](https://user-images.githubusercontent.com/102855127/182589386-3557790b-e28f-49e7-8060-a0eec2272db4.png)
![image](https://user-images.githubusercontent.com/102855127/182592748-9319231a-3116-4ced-af5e-727605a2bee4.png)


When I generated certificates, I generated only one set of certificates for both server and client. The hosts field was left blank because I did not want to limit the number of hosts.

Can I do this

**Reply by SimFG** [reply]
When adding a new node, generate the relevant key pair, and then start it according to the relevant configuration. There is a faster way, that is to generate key related files by etcd itself, see: auto-tls.

**Reply by oison111** [reply]
> When adding a new node, generate the relevant key pair, and then start it according to the relevant configuration. There is a faster way, that is to generate key related files by etcd itself, see: auto-tls.

I didn't add a new node, I just used a set of key files for both the server and the client, and I still reported the same error,
![image](https://user-images.githubusercontent.com/102855127/182594933-6f679fb5-bd01-4cbd-8c38-940c5b07d49a.png)

**Reply by SimFG** [reply]
you want client-server tls or peer-peer tls or both? peer and client are different. Now, you configure the same.

**Reply by oison111** [reply]
> you want client-server tls or peer-peer tls or both? peer and client are different. Now, you configure the same.

![image](https://user-images.githubusercontent.com/102855127/182601398-8ebf8e5e-2e45-461b-9c68-a0e47564d598.png)
It's one of two, right ??

**Reply by SimFG** [reply]
yes, `client-transport-security` is client to etcd server, and `peer-transport-security` is used between etcd nodes in etcd server.

**Reply by oison111** [reply]
> yes, `client-transport-security` is client to etcd server, and `peer-transport-security` is used between etcd nodes in etcd server.

**I specify the IP address in hosts**
![image](https://user-images.githubusercontent.com/102855127/182824221-9479ca06-b774-48ac-beea-33958bc776ec.png)

![image](https://user-images.githubusercontent.com/102855127/182824192-95cfe887-0df1-4e93-8cce-b88b9adc15fb.png)

**This enables you to start ETCD**
![image](https://user-images.githubusercontent.com/102855127/182824536-40322e20-878c-484a-aade-ee3c27551888.png)

**But this is not what I want, I don't want to specify a specific IP, I want any server to be able to join ETCD**
**just like this****
![image](https://user-images.githubusercontent.com/102855127/182824939-12f9cdfd-d1bb-45c3-87aa-106b24996240.png)

**However, it cannot be started, and the error will be "TLS bad certificate".**
![image](https://user-images.githubusercontent.com/102855127/182825228-c75eeef6-b517-4744-b9af-f2a50a56d58e.png)

**I refer to this link**
https://github.com/coreos/docs/blob/master/os/generate-self-signed-certificates.md
 
![image](https://user-images.githubusercontent.com/102855127/182825983-7366decb-eaf1-4a28-bf25-5f6fb8083943.png)

**Reply by SimFG** [reply]
"TLS bad certificate", does it happen when the etcd server start or when the client connects the etcd server?

**Reply by oison111** [reply]
> "TLS bad certificate", does it happen when the etcd server start or when the client connects the etcd server?

**When I run the following command on three machines at once** 
etcd --config-file /etc/etcd/etcd.config.yml

**Reply by oison111** [reply]
> "TLS bad certificate", does it happen when the etcd server start or when the client connects the etcd server?

**I've tried both of these, and they both give me the same error**
![image](https://user-images.githubusercontent.com/102855127/182846347-a87995dd-5dd9-476a-96f3-5a803dedf416.png)
![image](https://user-images.githubusercontent.com/102855127/182847219-bdf33db3-dbc4-45cc-95c3-a63bb4750664.png)
![image](https://user-images.githubusercontent.com/102855127/182847253-76039103-3c8f-4a83-b561-5a50ec78cb39.png)
![image](https://user-images.githubusercontent.com/102855127/182847285-ac380990-7f01-4ce0-ac33-6643ce7f35d2.png)

**Reply by SimFG** [reply]
I think you can understand the role of this json. There should be two json files, like ca-csr.json and peer.json.

**Reply by oison111** [reply]
> I think you can understand the role of this json. There should be two json files, like ca-csr.json and peer.json.
![image](https://user-images.githubusercontent.com/102855127/182851197-aaf5a069-78aa-4c90-afc2-9bffbce21850.png)
